Implement Calculator in Haskell

gabebw opened the pull request.

gabebw committed:
Initialize a new Haskell project

gabebw committed:
Parse arbitrary number of tokens

gabebw committed:
Make operator a type

gabebw committed:
Move types into their own module

gabebw committed:
Only expose the function we want

gabebw committed:
Add shunting-yards algorithm in Haskell

gabebw committed:
Return only what we want

gabebw committed:
Add the operators

gabebw committed:
OMG it outputs RPN

gabebw committed:
Add interpreter - it works!

gabebw committed:
`Node a` can just be `Node`

We don't do anything with the `a`, we always know what it's going to be.

gabebw committed:
Add a type alias for readability

jferris commented on haskell/src/Parser.hs:
```diff
@@ -0,0 +1,43 @@
+numberParser :: Parser String
+numberParser = integerParser
+
+operatorParser :: Parser Char
```
You can combine `operatorParser` and `operatorFromChar` in a way that doesn't require using `error` using `<|>`. I think it's something like:

``` haskell
operatorParser :: Parser Char
operatorParser = try (char '+' >> return Plus)
    <|> try (char '-' >> return Minus)
    <|> (char '*' >> return Times)
```

This will either map one of the matched characters to an enum value or result in a parser error.

jferris commented on haskell/src/Parser.hs:
```diff
@@ -0,0 +1,43 @@
+integerParser = many digit
+
+numberParser :: Parser String
+numberParser = integerParser
```
If you define this as `fmap read integerParser`, you'll get a `Parser Int` which is probably more useful. You could also use `fmap (NumberNode . read) integerParser` to get yourself a `Parser NumberNode`.

Parsers which return the type you want are usually easier to compose.

jferris commented on haskell/src/Parser.hs:
```diff
@@ -0,0 +1,43 @@
+    return $ [NumberNode (read number)] ++ (concat nodes)
+
+furtherExpressionParser :: Parser [Node]
+furtherExpressionParser = do
```
If you can define `operatorParser :: Parser Node` and `numberParser :: Node`, then you can define this more easily:

``` haskell
furtherExpressionParser :: Parser [Node]
furtherExpressionParser = sequence [numberParser, operatorParser]
```

Sequence will go from `[Parser Node]` to `Parser [Node]`.

gabebw committed:
Parse into the types we want earlier

Instead of `Parser String`, return `Parser Float`, `Parser NumberNode`, `Parser
Operator`, etc.

gabebw committed:
Test that Haskell parser is whitespace-insensitive

gabebw committed:
xNode parsers can consume all of their own input

For example, `numberNodeParser` consumes ` *\d+` instead of just `\d+`.

And the operator parser now returns a fully-formed OperatorNode.

gabebw committed:
sequence is so cool

gabebw committed:
Extract a list of all operator chars

gabebw committed:
Remove unused parameters

gabebw committed:
Use lookahead and combinators

This means that if we're looking for an operator and don't see "+", "-", or "*",
a parser error will bubble all the way up for us, instead of having to pattern
match ourselves and raise an error if we get unexpected input.

gabebw committed:
We may not have a result to give

gabebw committed:
Use many1 instead of many

many1 is "1 or more", while many is "0 or more"

gabebw committed:
Add an ApplicativeParser

`Parser` is an instance of Applicative, so we can use `<$>` (fmap), `<*>` (like
`(Just (5 +)) <*> (Just 6) == Just 11`), and `*>`, which is `<*>` but discarding
the left argument. It Just Works for each parser combination.

gabebw committed:
Rename Parser -> MonadicParser

gabebw committed:
Prefer <$> over fmap

gabebw committed:
Just be yourself
